File: src/main/java/fs/client/system/WaterSystem.java
package fs.client.system;

import fs.client.event.WorldGeneratedEvent;
import fs.client.event.TickEvent;
import fs.client.event.WaterUpdatedEvent;
import fs.client.world.BlockType;
import fs.client.world.Location;
import fs.client.world.World;

import javax.enterprise.event.Event;
import javax.enterprise.event.Observes;
import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.*;

import static fs.client.world.World.MAX_WATER_LEVEL;
import static java.util.Comparator.comparingInt;

@Singleton
public class WaterSystem {

  private static final int FLOW_INTERVAL = 1;

  @Inject
  private Event<Object> events;

  @Inject
  private World world;

  private int[] maxFlowOut;
  private boolean[] considered;

  private int tick = 0;
  private static Comparator<Location> WATER_POTENTIAL_COMPARATOR = comparingInt(WaterSystem::waterPotential);

  public void onWorldGenerated(@Observes WorldGeneratedEvent event) {
    maxFlowOut = new int[world.size()];
    considered = new boolean[world.size()];
  }

  public void onUpdateRequested(@Observes TickEvent event) {
    tick++;
    if (tick % FLOW_INTERVAL == 0) {
      flow();
    }
  }


  private boolean flow() {
    // Reset flow
    Arrays.fill(maxFlowOut, 0);
    Arrays.fill(considered, false);

    // Get all potential flow candidates
    List<Location> unsortedCandidates = new ArrayList<>();
    for (Location location: world.locations()) {
      if (isWater(location)) {
        if (!considered[location.index()]) {
          unsortedCandidates.add(location);
          considered[location.index()] = true;
        }

        for (Location neighbour: location.neighbours().values()) {
          if (neighbour.block().type() == BlockType.AIR) {
            if (!considered[neighbour.index()]) {
              unsortedCandidates.add(neighbour);
            }
          }
        }
      }
    }

    PriorityQueue<Location> flowCandidates = new PriorityQueue<>(WATER_POTENTIAL_COMPARATOR);
    Collections.shuffle(unsortedCandidates);
    flowCandidates.addAll(unsortedCandidates);

    // Process water
    boolean waterChanged = false;
    while (!flowCandidates.isEmpty()) {
      // Find path from water source to water target
      Location targetLocation = flowCandidates.poll();
      int targetWaterLevel = targetLocation.block().waterLevel();
      if (targetWaterLevel >= MAX_WATER_LEVEL) {
        continue;
      }

      List<Location> path = find(targetLocation);
      if (!path.isEmpty()) {
        for (int i = 0; i < path.size() - 1; i++) {
          maxFlowOut[path.get(i).index()]++;
        }

        Location sourceLocation = path.get(0);
        if (sourceLocation != null) {
          int sourceWaterLevel = sourceLocation.block().waterLevel();
          sourceLocation.block().waterLevel(sourceWaterLevel - 1);
          targetLocation.block().waterLevel(targetWaterLevel + 1);
          waterChanged = true;
          events.fire(new WaterUpdatedEvent(sourceLocation));
          events.fire(new WaterUpdatedEvent(targetLocation));


          flowCandidates.remove(sourceLocation);
          flowCandidates.add(sourceLocation);
          if (targetWaterLevel != MAX_WATER_LEVEL) {
            flowCandidates.add(targetLocation);
          }
        }
      }
    }

    return waterChanged;
  }

  private List<Location> find(Location startLocation) {
    int targetWaterPotential = waterPotential(startLocation) + 2;

    Map<Location, Location> cameFrom = new HashMap<>();
    cameFrom.put(startLocation, null);

    Queue<Location> unvisitedCells = new LinkedList<>();
    unvisitedCells.add(startLocation);
    while (!unvisitedCells.isEmpty()) {
      Location cellLocation = unvisitedCells.poll();

      // Find cell with higher potential
      int cellPotential = waterPotential(cellLocation);
      if (cellPotential >= targetWaterPotential) {
        List<Location> path = new ArrayList<>();
        path.add(cellLocation);
        while (cameFrom.get(cellLocation) != null) {
          cellLocation = cameFrom.get(cellLocation);
          path.add(cellLocation);
        }

        return path;
      }

      // If not higher potential, findComponentAt neighbours
      for (Location neighbour : cellLocation.neighbours().values()) {
        if (maxFlowOut[neighbour.index()] < neighbour.block().waterLevel()) {
          if (!cameFrom.containsKey(neighbour)) {
            unvisitedCells.add(neighbour);
            cameFrom.put(neighbour, cellLocation);
          }
        }
      }
    }

    // If we cannot findComponentAt any path, then any adjacent neighbours would not findComponentAt any better solution either; discard
    // them so we don't waste time!
    for (Location cellLocation : cameFrom.keySet()) {
      maxFlowOut[cellLocation.index()] = cellLocation.block().waterLevel();
    }

    return Collections.emptyList();
  }

  private static int waterPotential(Location location) {
    int waterLevel = location.block().waterLevel();
    return location.y() * (MAX_WATER_LEVEL + 1) + waterLevel;
  }

  public boolean isWater(Location location) {
    return location.block().waterLevel() > 0;
  }
}

File: src/main/kotlin/com/fracturedskies/water/WaterSystem.kt
package com.fracturedskies.water

import com.fracturedskies.api.*
import com.fracturedskies.api.block.data.WaterLevel
import com.fracturedskies.engine.api.*
import com.fracturedskies.engine.collections.forEach
import com.fracturedskies.engine.math.*
import com.fracturedskies.engine.math.Vector3i.Companion.AXIS_Y
import com.fracturedskies.engine.math.Vector3i.Companion.xZSpiral
import com.fracturedskies.water.WaterMap.Companion.MAX_WATER_RANGE
import java.lang.Integer.min
import java.util.*
import java.util.Comparator.comparingInt
import java.util.function.ToIntFunction
import javax.enterprise.event.Observes
import javax.inject.*

@Singleton
class WaterSystem {

  @Inject
  lateinit var world: World

  private var initialized = false
  private lateinit var water: WaterMap
  private lateinit var pathFinder: PathFinder

  fun onNewGameRequested(@Observes newGameRequested: NewGameRequested) {
    if (!initialized) {
      water = WaterMap(world.dimension)
      pathFinder = PathFinder({ _, toPos -> water.has(toPos) && water.maxFlowOut[toPos] != 0.toByte() })
      initialized = true
    }
  }

  fun onWorldGenerated(@Observes worldGenerated: WorldGenerated) {
    if (!initialized) {
      water = WaterMap(world.dimension)
      pathFinder = PathFinder({ _, toPos -> water.has(toPos) && water.maxFlowOut[toPos] != 0.toByte() })
      initialized = true
    }

    val waterLevelUpdates = mutableMapOf<Vector3i, Byte>()
    worldGenerated.blocks.forEach { pos, block ->
      water.setOpaque(pos, block.type.opaque)
      if (block.type.opaque && water.getLevel(pos) != 0.toByte()) {
        water.setLevel(pos, 0)
        waterLevelUpdates[pos] = 0
      }
    }
    worldGenerated.blocks.forEach { pos, _ ->
      water.nearestWaterDrop[pos] = calculateNearestWaterDrop(pos)
    }
    if (waterLevelUpdates.isNotEmpty()) {
      world.updateBlocks(waterLevelUpdates
          .map { it.key to world.blocks[it.key].with(WaterLevel(it.value)) }
          .toMap(), Cause.of(this))
    }
  }

  fun onBlockUpdated(@Observes blocksUpdated: BlocksUpdated) {
    if (!initialized) {
      water = WaterMap(world.dimension)
      pathFinder = PathFinder({ _, toPos -> water.has(toPos) && water.maxFlowOut[toPos] != 0.toByte() })
      initialized = true
    }

    val waterLevelUpdates = mutableMapOf<Vector3i, Byte>()
    blocksUpdated.blocks.forEach { (pos, block) ->
      water.setOpaque(pos, block.type.opaque)
      water.setLevel(pos, block[WaterLevel::class]!!.value)
      if (block.type.opaque && water.getLevel(pos) != 0.toByte()) {
        water.setLevel(pos, 0)
        waterLevelUpdates[pos] = 0
      }
    }
    blocksUpdated.blocks.forEach { (pos, _) ->
      updateNearestWater(pos)
    }
    if (waterLevelUpdates.isNotEmpty()) {
      world.updateBlocks(waterLevelUpdates
          .map { it.key to world.blocks[it.key].with(WaterLevel(it.value)) }
          .toMap(), Cause.of(this))
    }
  }

  fun onUpdate(@Observes update: Update) {
    val flow = flow()
    val evaporation = evaporation()

    val updates = flow + evaporation
    if (updates.isNotEmpty()) {
      world.updateBlocks(updates
          .map { it.key to world.blocks[it.key].with(WaterLevel(it.value)) }
          .toMap(), Cause.of(this))
    }
  }

  private fun updateNearestWater(pos: Vector3i) {
    xZSpiral(MAX_WATER_RANGE)
        .flatMap { listOf(pos + AXIS_Y + it, pos + it) }
        .filter { water.has(it) }
        .forEach {
          val newValue = calculateNearestWaterDrop(it)
          if (water.nearestWaterDrop[it] != newValue) {
            water.nearestWaterDrop[it] = newValue
            water.sea[it]?.disturbed = true
          }
        }
  }

  private fun calculateNearestWaterDrop(pos: Vector3i): Int {
    if (water.getOpaque(pos)) return WaterMap.MAX_WATER_RANGE
    return xZSpiral(MAX_WATER_RANGE).map { pos - AXIS_Y + it }
        .filter { water.has(it) }
        .filter { ! water.getOpaque(it) }
        .map { min(it distanceTo (pos - Vector3i.AXIS_Y), MAX_WATER_RANGE) }
        .min() ?: MAX_WATER_RANGE
  }

  private fun evaporation(): Map<Vector3i, Byte> {
    val evaporation = water.evaporationCandidates
            .filter { Math.random() > 0.99995 }
    return if (evaporation.isNotEmpty()) {
      val waterLevelUpdates = mutableMapOf<Vector3i, Byte>()
      for (pos in evaporation) {
        water.setLevel(pos, 0)
        waterLevelUpdates[pos] = 0
      }
      waterLevelUpdates
    } else {
      mapOf()
    }
  }

  private fun flow(): Map<Vector3i, Byte> {
    water.maxFlowOut.clear()
    val originalWaterLevel = mutableMapOf<Vector3i, Byte>()
    val waterLevelUpdates = mutableMapOf<Vector3i, Byte>()
    val disturbedSeas = water.seas.filter { it.disturbed }
    disturbedSeas.forEach { it.disturbed = false }
    val providences = disturbedSeas.map {it.providence.toList()}
    for (providence in providences) {
      // Get all potential flow candidates
      for (adjacent in providence)
        water.maxFlowOut[adjacent] = water.getLevel(adjacent)

      Collections.shuffle(providence)
      val candidates = PriorityQueue<Vector3i>(waterPotentialComparator)
      candidates.addAll(providence)

      // Process water
      while (!candidates.isEmpty()) {
        // Find path from water source to water target
        val targetLocation = candidates.poll()
        val targetWaterLevel = water.getLevel(targetLocation)
        if (targetWaterLevel >= MAX_WATER_LEVEL) {
          continue
        }

        val targetWaterPotential = waterPotential(targetLocation) + 2
        val results = pathFinder.find(targetLocation, { waterPotential(it) >= targetWaterPotential}, { _, it -> -waterPotential(it)}, true)
        val path = results.path
        if (!path.isEmpty()) {
          path.stream().skip(1).forEach { water.maxFlowOut[it]-- }

          val sourceLocation = path.last()
          val sourceWaterLevel = water.getLevel(sourceLocation)

          originalWaterLevel.putIfAbsent(sourceLocation, sourceWaterLevel)
          originalWaterLevel.putIfAbsent(targetLocation, targetWaterLevel)
          val newSourceWaterLevel = sourceWaterLevel.dec()
          val newTargetWaterLevel = targetWaterLevel.inc()
          water.setLevel(sourceLocation, newSourceWaterLevel)
          water.setLevel(targetLocation, newTargetWaterLevel)
          waterLevelUpdates[sourceLocation] = newSourceWaterLevel
          waterLevelUpdates[targetLocation] = newTargetWaterLevel

          candidates.remove(sourceLocation)
          candidates.add(sourceLocation)
          if (targetWaterLevel != MAX_WATER_LEVEL)
            candidates.add(targetLocation)
        } else {
          for (entry in results.cameFrom) {
            val to = entry.key
            var from = entry.value
            if (from != null) {
              val dir = when {
                to.x > from.x -> Vector3i.AXIS_X
                to.x < from.x -> Vector3i.AXIS_NEG_X
                to.y > from.y -> AXIS_Y
                to.y < from.y -> Vector3i.AXIS_NEG_Y
                to.z > from.z -> Vector3i.AXIS_Z
                else -> Vector3i.AXIS_NEG_Z
              }

              while (from != to) {
                water.maxFlowOut[from] = 0
                from += dir
              }
            }
            water.maxFlowOut[to] = 0
          }
        }
      }
    }

    // Split seas that have settled
    disturbedSeas.filterNot { it.disturbed }.forEach { water.splitSea(it) }

    // Only update water levels for actual change (rather than pass-through changes)
    return waterLevelUpdates.filter { (pos, waterLevel) -> originalWaterLevel[pos] != waterLevel }
  }

  private fun waterPotential(pos: Vector3i) = pos.y * (MAX_WATER_LEVEL + MAX_WATER_RANGE + 1) + water.nearestWaterDrop[pos] + water.getLevel(pos)

  private val waterPotentialComparator = comparingInt(ToIntFunction<Vector3i> { waterPotential(it) })
}
